src/tqec/circuit/detector/construction.py

```diff
@@ -0,0 +1,408 @@
+    )
+
+
+def _compile_fragment_loop_into_circuit(
+    fragment_loop: FragmentLoop,
+    state: State,
+    out_circuit: stim.Circuit,
+    qubit_coords_map: dict[int, list[float]],
+) -> None:
+    if not fragment_loop.have_detector_sources and not state.end_stabilizers:
+        body_circuit = compile_fragments_to_circuit(
+            fragment_loop.fragments, qubit_coords_map
+        )
+        out_circuit += body_circuit * fragment_loop.repetitions
+    elif fragment_loop.repetitions == 1:
+        _compile_fragments_into_circuit(
+            fragment_loop.fragments,
+            state,
+            out_circuit,
+            qubit_coords_map,
+        )
+    else:
+        # compile for the first repetition to account for
+        # possible boundary effects
+        first_loop_circuit = stim.Circuit()
+        _compile_fragment_loop_into_circuit(
+            fragment_loop.with_repetitions(1),
+            state,
+            first_loop_circuit,
+            qubit_coords_map,
+        )
+        second_loop_circuit = stim.Circuit()
+        _compile_fragment_loop_into_circuit(
+            fragment_loop.with_repetitions(1),
+            state,
+            second_loop_circuit,
+            qubit_coords_map,
+        )
+        if first_loop_circuit == second_loop_circuit:
+            out_circuit += first_loop_circuit * fragment_loop.repetitions
+        else:
+            out_circuit += first_loop_circuit
+            out_circuit += second_loop_circuit * (fragment_loop.repetitions - 1)
+
+
+def construct_boundary_stabilizers(
+    sources: list[PauliString],
+    circuit: stim.Circuit,
+    qubit_coords_map: dict[int, list[float]],
+    inverse: bool = False,
+    boundary_collapse: list[PauliString] = (),
+) -> list[BoundaryStabilizer]:
+    # since the circuit is from a fragment, it's safe to ignore collapses
+    circuit_tableau = circuit.to_tableau(
+        ignore_noise=True, ignore_measurement=True, ignore_reset=True
+    )
+    if inverse:
+        circuit_tableau = circuit_tableau.inverse(unsigned=True)
+    stabilizers = []
+    for source_index, source in enumerate(sources):
+        before_collapse = (
+            source.after(circuit_tableau, targets=range(circuit.num_qubits))
+            if len(circuit_tableau) > 0
+            else source
+        )
+        after_collapse = before_collapse.collapse_by(boundary_collapse)
+        commute_collapse_indices = []
+        anticommute_collapse_indices = []
+        for i, collapse in enumerate(boundary_collapse):
+            index = i - len(boundary_collapse)
+            if not collapse.intersects(before_collapse):
+                continue
+            if before_collapse.commutes(collapse):
+                commute_collapse_indices.append(index)
+            else:
+                anticommute_collapse_indices.append(index)
+        if inverse:
+            if source.weight == 0:
+                source_measurement_indices = frozenset()
```

Comment:
Is that branch even possible? Looking for the usages of this function, the `sources` argument is given:

- `fragment.begin_stabilizer_sources`
- `state.resets_passed_in + fragment.end_stabilizer_sources`

It seems like, in both cases, the `PauliString` should be at least weight 1 as they originate from a reset/measurement. Do I miss something?

Reply:
At the first matching interface(the beginning boundary of the first fragment), the stabilizers from the measurements at the end of the fragment will match to the "I" stabilizers intentionally made for each qubit as the end stabilizers from the virtual previous fragment:
```python
def compile_fragments_to_circuit(
    fragments: Iterable[Fragment | FragmentLoop],
    qubit_coords_map: dict[int, list[float]],
) -> stim.Circuit:
    all_qubits = sorted(qubit_coords_map.keys())
    state = State(
        resets_passed_in=list(),
        previous_measurements=list(),
        end_stabilizers=[
            BoundaryStabilizer(
                before_collapse=PauliString({q: "I"}),
                after_collapse=PauliString({q: "I"}),
                coords=tuple(qubit_coords_map[q]),
            )
            for q in all_qubits
        ], # Here are the "I" stabilizers
    )
    out_circuit = stim.Circuit()
    _compile_fragments_into_circuit(
        fragments, state, out_circuit, qubit_coords_map, True
    )
    return out_circuit
```

The same situation is for the end of the circuit(the ending boundary of the last fragment):

```python
    if compile_final_state_once_more:
        _compile_fragment_into_circuit(
            Fragment(
                circuit=stim.Circuit(),
                begin_stabilizer_sources=[
                    PauliString({q: "I"}) for q in sorted(qubit_coords_map.keys())
                ],
            ),
            state,
            out_circuit,
            qubit_coords_map,
        )
```